Track recent reward balances & implemented lagged reward balance requests

jagerman opened the pull request:
This adds a batching rewards table to always keep around the most recent 5 blocks (in addition to current block), and then uses that to implement rewards signing requests that request the reward at a recent height, rather than relying on each SN to use its own current height.

The RPC endpoint now takes a "height" parameter, which can be a <= 0 value to indicate "N blocks ago" (so 0 = most recent, -1 = one block ago).  The default, if omitted, is -1 so that requests get lagged by a block by default, which should avoid issues with a request being made at the same time a block is propagating through the network.

jagerman committed:
Fix assertion failure on mainnet debug build syncing

jagerman committed:
Drop unused table; abstract table existance query

jagerman committed:
Track recent blocks' accrued rewards

This will allow us to certify the earned amount a block or two behind
the tip to not have to worry about timing the request right so that
"current height" won't differ across the network during the distributed
balance signing request.

jagerman committed:
Add height to RPC and OMQ reward request endpoints

- Allows passing a relative height (0, -1, -2) to the RPC endpoint to
  indicate "current top block", "one block ago", etc.  The default is
  -1, which should solve the race condition with a new block arriving.

- Reformulates the request data for the rewards request to be a bt-dict
  with address and height, and updates the request to use the given
  height rather than current chain height.

Doy-lee commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -426,22 +461,55 @@ static std::pair<uint64_t, uint64_t> get_accrued_rewards_impl(
+    if (!rewards) {
+        // No rewards found; check to see if we actually have any recent records for that height and
+        // if not, return a "don't know" nullopt value.  Otherwise we fall through and return an
+        // authoritive 0 value.
```
Why 0? In the bls aggregator usage code that 0 is going to trigger the "invalid request because you have earnt 0 SENT" which is going to be misleading especially if you're targeting the tip of the chain after you've just received a block.

I think more appropriate is returning another std::nullopt, or, pull the maximum known rewards value (which would be the most correct output if you pass in a much higher height than the DB has)

jagerman replied:
Good point; pushed an update.  (We don't have to actually need to query that one as we have a `height` member already tracking it).

jagerman replied:
For heights that are in between min-height-we-have and curr-height we still do want to return 0, because that does indeed mean that address has no SENT earned.

jagerman committed:
Fix reward return values of >= current height

Return nullopt if requesting a balance at a height higher than what we
have, and use the version that fetches from our current rewards table if
asked for reward at the current height.
